pkg/kv/txn_coord_sender.go

```diff
@@ -1086,6 +1086,10 @@ func (tc *TxnCoordSender) heartbeat(ctx context.Context) bool {
 		return false
 	}
 
+	if txn.Key == nil {
```

Comment:
May want to add a comment here explaining why this might be nil:

```
  // txn.Key may be nil here if the batch containing the `BeginTransaction` has not
  // yet returned successfully. Just skip the heartbeat in this case and wait for the
  // next heartbeat loop tick.
```